L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,1095 @@
+  HGCalModuleDetId hgc_module_id(module_id);
+  unsigned subdet = hgc_module_id.triggerSubdetId();
+
+  geom_ordered_set trigger_cell_det_ids;
+
+  // Scintillator
+  if (subdet == HGCalTriggerSubdetector::HGCalHScTrigger) {
+    int ieta0 = hgc_module_id.eta() * hSc_module_size_;
+    int iphi0 = (hgc_module_id.phi() * (hgc_module_id.sector() + 1)) * hSc_module_size_;
+
+    for (int ietaAbs = ieta0; ietaAbs < ieta0 + (int)hSc_module_size_; ietaAbs++) {
+      int ieta = ietaAbs * hgc_module_id.zside();
+      for (int iphi = iphi0; iphi < iphi0 + (int)hSc_module_size_; iphi++) {
+        unsigned trigger_cell_id = HGCScintillatorDetId(hgc_module_id.type(), hgc_module_id.layer(), ieta, iphi);
+        if (validTriggerCellFromCells(trigger_cell_id))
+          trigger_cell_det_ids.emplace(trigger_cell_id);
+      }
+    }
+  }
+
+  // HFNose
+  else if (subdet == HGCalTriggerSubdetector::HFNoseTrigger) {
+    HFNoseDetId module_nose_id(module_id);
+    HFNoseDetIdToModule hfn;
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (auto const& idx : ids) {
+      if (validTriggerCellFromCells(idx.rawId()))
+        trigger_cell_det_ids.emplace(idx);
+    }
+  }
+  // Silicon
+  else {
+    HGCSiliconDetIdToROC tc2roc;
+    int moduleU = hgc_module_id.moduleU();
+    int moduleV = hgc_module_id.moduleU();
+    unsigned layer = hgc_module_id.layer();
+
+    //Rotate to sector
+    geom_rotation_120_.uvMappingFromSector0(getWaferCentring(layer, subdet), moduleU, moduleV, hgc_module_id.sector());
+
+    DetId::Detector det = (subdet == HGCalTriggerSubdetector::HGCalEETrigger ? DetId::HGCalEE : DetId::HGCalHSi);
+
+    unsigned wafer_type = detIdWaferType(det, layer, moduleU, moduleV);
+    int nroc = (wafer_type == HGCSiliconDetId::HGCalFine ? 6 : 3);
+    // Loop on ROCs in wafer
+    for (int roc = 1; roc <= nroc; roc++) {
+      // loop on TCs in ROC
+      auto tc_uvs = tc2roc.getTriggerId(roc, wafer_type);
+      for (const auto& tc_uv : tc_uvs) {
+        HGCalTriggerDetId trigger_cell_id(
+            subdet, hgc_module_id.zside(), wafer_type, layer, moduleU, moduleV, tc_uv.first, tc_uv.second);
+        if (validTriggerCellFromCells(trigger_cell_id.rawId()))
+          trigger_cell_det_ids.emplace(trigger_cell_id);
+      }
+    }
+  }
+
+  return trigger_cell_det_ids;
+}
+
+HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp3::getNeighborsFromTriggerCell(
+    const unsigned trigger_cell_id) const {
+  throw cms::Exception("FeatureNotImplemented") << "Neighbor search is not implemented in HGCalTriggerGeometryV9Imp3";
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::getLinksInModule(const unsigned module_id) const {
+  HGCalModuleDetId module_det_id(module_id);
+  unsigned subdet = module_det_id.triggerSubdetId();
+
+  unsigned links = 0;
+  // Scintillator
+  if (subdet == HGCalTriggerSubdetector::HFNoseTrigger) {
+    links = 1;
+  }
+  // TO ADD HFNOSE : getLinksInModule
+  // Silicon
```

Comment:
Comment to be changed: this is for both silicon and scintillator in HGC

Reply:
done